functionalDB/chapter.md

```diff
@@ -0,0 +1,1138 @@
+   ([id] (Entity.  id {})))
+````
+Note that if no id is given, the entity’s id is set to be *:db/no-id-yet*, which means that something else is responsible for giving it an id. We’ll see how that works later.
+
+**Attributes**
+
+Each attribute consists of its name, value, and the timestamps of both its most recent and previous update. In addition to these fields, each attribute has two fields that describe its *type* and *cardinality*.
+
+In the case that an attribute is used to represent a relationship to another entity, its *type* will be *:db/ref* and its value will be the id of the related entity. This simple type system also acts as an extension point. Users are free to define their own types and leverage them to provide additional semantics for their data.
+
+An attribute's *cardinality* specifies whether the attribute represents a single value or a set of values. We use this field to determine the set of operations that are permitted on this attribute.
+
+Similar to entities, creating an attribute is done using the *make-attr* function 
+
+````clojure
+(defrecord Attr [name value ts prev-ts])
+
+(defn make-attr
+   ([name value type ; these ones are required
+       & {:keys [cardinality] :or {cardinality :db/single}} ]
+     {:pre [(contains? #{:db/single :db/multiple} cardinality)]}
+    (with-meta (Attr. name value -1 -1) {:type type :cardinality cardinality})))
+````
+There are a couple of interesting patterns used in this constructor function: 
+
+* We use Clojure’s _Design by Contract_ [ADD REF HERE] pattern to validate that the cardinality parameter is a permissible value
+* We use Clojure’s destructuring mechanism to provide a default value of *:db/single* if one is not given
+* We use Clojure’s metadata capabilities to separate between an attribute's data (name, value and timestamps) and its metadata (type and cardinality). In Clojure, metadata handling is done using the *with-meta* (setting metadata) and *meta* (reading metadata) functions.
+
+Attributes only have meaning if they are part of an entity. We make this connection with the *add-attr* function, which adds a given attribute to an entity's attribute map (called *:attrs*). 
+
+Note that instead of directly using the attribute’s name, we first convert it into a keyword to adhere to Clojure’s idiomatic usage of maps.
+
+````clojure
+(defn add-attr [ent attr]
+   (let [attr-id (keyword (:name attr))]
+      (assoc-in ent [:attrs attr-id] attr)))
+````
+
+### Storage
+
+So far, we have talked a lot about _what_ we are going to store, without thinking about _where_ we are going to store it. In this chapter, we resort to the simplest storage mechanism, which is storing the data in memory. This is certainly not reliable, but it simplifies development and debugging and allows us to focus on more interesting parts of the program. 
+
+We will access the storage via a simple _protocol_ that will make it possible to add more durable storage providers in the future.

```

Comment:
The "add more durable storage providers in the future" part threw me a bit.  Does this mean "switch to a more durable storage provider in the future", or can there be multiple storage providers at once?